### PR TITLE
Sigma message construction and parsing for CASE

### LINKED ISSUE #4469: Implement Sigma message construction and parsing for CASE state machine
#### Problem
CASE state machine is barebones, and is missing Sigma message construction and parsing code.
 
 #### Proposed Solution
Implement Sigma message constructors and parser and use in CASE state machine.

### PR DESCRIPTION
**Pull request by jpk233**
#### Problem
CASE session functionality, along with sigma message protocol, is not implemented. Additionally, an operational certificate class is lacking for API usage.

 #### Summary of Changes
 * CASE session support added, currently only for single device certificates.
 * Sigma message construction and parsing implemented.
 * CHIP Operational cert data types added 
 

 Fixes #4469 #1802

### CODE REVIEW

File: src/transport/CASESession.cpp [hunk 1]
```diff
@@ -54,15 +71,145 @@ void CASESession::Clear()
...
+exit:
+    return error;
+}
+
+CHIP_ERROR CASESession::Init(ChipOperationalCredentialSet * chipOperationalCredentialSet, NodeId myNodeId, uint16_t myKeyId,
+                             SessionEstablishmentDelegate * delegate)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    VerifyOrExit(delegate != nullptr, err = CHIP_ERROR_INVALID_ARGUMENT);
```
Comment (pan-apple) [inline src/transport/CASESession.cpp:178]: We can switch to using `ReturnMacros.h` instead of `VerifyOrExit()` and `SuccessOrExit()`. It'll reduce some boilerplate code. It specifically useful when the `exit:` is just returning error code (instead of doing some additional cleanup).

The code will change to `VerifyOrReturnError(delegate != nullptr, CHIP_ERROR_INVALID_ARGUMENT);`

The following code
```
     err = mCommissioningHash.Begin();
     SuccessOrExit(err);
```

will change to 
```
    ReturnErrorOnFailure(mCommissioningHash.Begin());
```

I understand that the current PR is replicating the code structure from `PASESession.cpp`. Some of that code predates `ReturnMacros` and could use a refresh as well.

File: src/credentials/CHIPOpCred.h [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+using namespace Crypto;
+
+/**
+ *  @class ChipOperationalCredentialSet
+ *
+ *  @brief
+ *    Collection of CHIPCertificateSet providing methods for
+ *    finding and validating Device Credentials against Roof of Trust chains.
+ */
+class DLL_EXPORT ChipOperationalCredentialSet
```
Comment (pan-apple) [inline src/credentials/CHIPOpCred.h:44]: Let's remove `Chip` from the class name. The namespace is already accounting for it.

File: src/credentials/CHIPOpCred.cpp [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+ *      This file defines data types and objects for modeling and
+ *      working with CHIP Operational Credentials.
+ *
+ */
+
+#ifndef __STDC_LIMIT_MACROS
+#define __STDC_LIMIT_MACROS
+#endif
+
+#include <credentials/CHIPOpCred.h>
```
Comment (pan-apple) [inline src/credentials/CHIPOpCred.cpp:30]: Should we rename the files to `CHIPOperationalCredentials.cpp/.h`?

**Review by pan-apple — COMMENTED** [review]
The code structure is great! I have some minor comments.
Did you get to test the code on a device to setup the CASE session?

**Comment by pan-apple** [comment]
Tag: @balducci-apple

File: src/credentials/CHIPOpCred.h (hunk 2 above)
Comment (andy31415) [reply src/credentials/CHIPOpCred.h:44]: Maybe even from the header name. The project is chip - although we inherited some naming starting with CHIP they seem redundant.

File: src/credentials/CHIPOpCred.h [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+     **/
+    ChipCertificateSet * FindCertSet(const CertificateKeyId & trustedRootId) const;
+
+    /**
+     * @return A pointer to the set of certificate-set data entries.
+     **/
+    const ChipCertificateSet * GetOpCredSet() const { return mOpCreds; }
+
+    /**
+     * @return A pointer to the last certificate-set in the set. Returns NULL if certificate set is empty.
```
Comment (andy31415) [inline src/credentials/CHIPOpCred.h:81]: What is the use for getting the last entry?

The restriction of 'may be null' seem to require extra logic in any users of the class.
I would suggest to remove this and then add it (likely modified without the null requirement) as a method when it is required.

File: src/credentials/CHIPOpCred.cpp [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+}
+
+CHIP_ERROR ChipOperationalCredentialSet::SignMsg(const uint8_t * msg, const size_t msg_length, P256ECDSASignature & out_signature)
+{
+    return mDeviceOpCredKeypair->ECDSA_sign_msg(msg, msg_length, out_signature);
+}
+
+const CertificateKeyId * ChipOperationalCredentialSet::GetTrustedRootId(uint16_t certSetIndex) const
+{
+    const ChipCertificateData * chipCertificateData = mOpCreds[certSetIndex].GetCertSet();
```
Comment (andy31415) [inline src/credentials/CHIPOpCred.cpp:155]: Could we validate certSetIndex validity?

File: src/credentials/CHIPOpCred.h [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+     * @brief Check whether Trusted Root ID is in the operational credential set.
+     *
+     * @param trustedRoot  Reference to the Trusted Root ID.
+     *
+     * @return True if Trusted Root ID is in the set, false otherwise.
+     **/
+    bool IsTrustedRootIn(const CertificateKeyId & trustedRoot) const;
+
+    /**
+     * @brief Validate CHIP certificate.
```
Comment (andy31415) [inline src/credentials/CHIPOpCred.h:118]: This @Brief restates the code: `ValidateCert` obviously validates the certificate (and we are in project chip, so 'CHIP certificate' also seems implied.

Could we:
  - replace abbreviations throughout: `ValidateCertificate` to favor readability over typing speed
  - add doccomments that state things not obvious. In this particular example, I wonder 'What is a valid cip certificate?'

File: src/credentials/CHIPOpCred.h [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+     * @param cert              Pointer to the CHIP certificiate to be validated. The certificate is
+     *                          required to be in this set, otherwise this function returns error.
+     * @param context           Certificate validation context.
+     *
+     * @return Returns a CHIP_ERROR on validation or other error, CHIP_NO_ERROR otherwise
+     **/
+    CHIP_ERROR ValidateCert(const CertificateKeyId & trustedRootId, const ChipCertificateData * cert, ValidationContext & context);
+
+    /**
+     * @brief Find and validate CHIP certificate.
```
Comment (andy31415) [inline src/credentials/CHIPOpCred.h:131]: brief seems to slightly mismatch with the name: I expect `FindValidCert` to "Find a valid certificate".

File: src/credentials/CHIPOpCred.h [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+     *
+     * @param trustedRootId    Reference to the Trusted Root ID for the Certificate Set to be used
+     *                          for validation
+     * @param subjectDN         Subject distinguished name to use as certificate search parameter.
+     * @param subjectKeyId      Subject key identifier to use as certificate search parameter.
+     * @param context           Certificate validation context.
+     * @param cert              A pointer to the valid CHIP certificate that matches search criteria.
+     *
+     * @return Returns a CHIP_ERROR on validation or other error, CHIP_NO_ERROR otherwise
```
Comment (andy31415) [inline src/credentials/CHIPOpCred.h:140]: how about 'Not found'? I imagine this would rather be 'returns an error if no valid certificate could be found'.

But this is obvious: a method returning CHIP_ERROR is expected to return error on failure and NO_ERROR on success. Can we say something non-obvious in the comments? Like is cert expected to be nullptr on error or is the return value undefined if error is returned?

File: src/credentials/CHIPOpCred.h [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+
+    /**
+     * @brief A function to sign a msg using ECDSA and the respective device credentials keypair.
+     *
+     * @param msg           Message that needs to be signed
+     * @param msg_length    Length of message
+     * @param out_signature Buffer that will hold the output signature. The signature consists of: 2 EC elements (r and s),
+     *                      represented as ASN.1 DER integers, plus the ASN.1 sequence Header
+     *
+     * @return Returns a CHIP_ERROR on error, CHIP_NO_ERROR otherwise
```
Comment (andy31415) [inline src/credentials/CHIPOpCred.h:153]: Please delete obvious doxygens. The take up vertical space when reading the header and do not seem to help in understanding code.

File: src/credentials/CHIPOpCred.h [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+     **/
+    const uint8_t * GetDevOpCred() const { return mDeviceCredentialsDER; }
+
+    /**
+     * @return Length of the loaded device credentials buffer.
+     **/
+    uint16_t GetDevOpCredLen() const { return mDeviceCredentialsLen; }
+
+private:
+    ChipCertificateSet * mOpCreds; /**< Pointer to an array of certificate data. */
```
Comment (andy31415) [inline src/credentials/CHIPOpCred.h:168]: What is the lifetime/ownership of these?

I see destrutor is empty, so presumably something else owns these items?
I find object ownership hard to follow when using these raw pointers.

File: src/transport/CASESession.h [hunk 11]
```diff
@@ -155,6 +221,14 @@ class DLL_EXPORT CASESession
 
     Protocols::SecureChannel::MsgType mNextExpectedMsg = Protocols::SecureChannel::MsgType::CASE_SigmaErr;
 
+    Hash_SHA256_stream mCommissioningHash;
```
Comment (andy31415) [inline src/transport/CASESession.h:224]: Are these used currently? I was wondering why the class needs an ongoing hash and extra data, but could not find usages.

Could we only add data members once used, so we can review usage and intent as well?

File: src/transport/CASESession.h [hunk 12]
```diff
@@ -147,6 +210,9 @@ class DLL_EXPORT CASESession
     void SendErrorMsg(SigmaErrorType errorCode);
     void HandleErrorMsg(const PacketHeader & header, const System::PacketBufferHandle & msg);
 
+    // TODO
```
Comment (andy31415) [inline src/transport/CASESession.h:213]: TODO?

File: src/transport/CASESession.h [hunk 13]
```diff
@@ -37,7 +39,29 @@
...
 using namespace Crypto;
+using namespace Credentials;
+
+struct CASESessionSerialized;
+
+struct CASESessionSerializable
+{
+    uint16_t mI2RR2IKeyLen;
+    uint8_t mI2RR2IKey[kAEADKeySize * 2];
+    uint8_t mPairingComplete;
```
Comment (andy31415) [inline src/transport/CASESession.h:59]: this is named like a bool, but stored as uint8_t.can it be a bool?

Can the items below be chip::NodeId (and some keyid for key indices) for clarity?

File: src/transport/CASESession.h [hunk 14]
```diff
@@ -126,6 +158,30 @@ class DLL_EXPORT CASESession
 
     Transport::PeerConnectionState & PeerConnection() { return mConnectionState; }
 
+    /** @brief Serialize the Pairing Session to a string.
```
Comment (andy31415) [inline src/transport/CASESession.h:161]: Does not seem to serialize to a 'string' but rather to serialized output.

Could we drop non-obvious doxygens? Return error on error and no-error on success is obvious and so is the fact that 'Serialize serializes things'.

File: src/transport/CASESession.cpp [hunk 15]
```diff
@@ -121,81 +270,287 @@ CHIP_ERROR CASESession::DeriveSecureSession(SecureSession & session)
...
+        err = mCommissioningHash.Finish(md);
+        SuccessOrExit(err);
+        bbuf.Put(md, kSHA256_Hash_Length);
+
+        VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_NO_MEMORY);
+    }
+
+    msg_salt->SetDataLength(saltlen);
+
+    err = HKDF_SHA256(&*mSharedSecret, mSharedSecret.Length(), msg_salt->Start(), saltlen, kKDFSR2Info, kKDFInfoLength, sr2k,
```
Comment (andy31415) [inline src/transport/CASESession.cpp:478]: &* looks odd. Maybe some Ptr() method would be easier to read.

File: src/transport/CASESession.cpp [hunk 16]
```diff
@@ -217,47 +572,353 @@ CHIP_ERROR CASESession::HandleSigmaR2_and_SendSigmaR3(const PacketHeader & heade
...
+    err = SendSigmaR3();
+    SuccessOrExit(err);
 
-    System::PacketBufferHandle msg_R3;
+exit:
+    return err;
+}
+
+CHIP_ERROR CASESession::HandleSigmaR2(const PacketHeader & header, const System::PacketBufferHandle & msg)
+{
```
Comment (andy31415) [inline src/transport/CASESession.cpp:586]: This is a 200-line method with 15 local variables and 2 arguments. Could we split it into something more manegeable?
Could we use some RAII to avoid the need for goto exit and then not be forced to declare all variables at top (somewhat better scoping to help readability)?

File: src/transport/CASESession.cpp [hunk 17]
```diff
@@ -121,81 +270,287 @@ CHIP_ERROR CASESession::DeriveSecureSession(SecureSession & session)
...
+    }
+    else if (err != CHIP_NO_ERROR)
+    {
+        SendErrorMsg(SigmaErrorType::kUnexpected);
+    }
+    return err;
+}
+
+CHIP_ERROR CASESession::SendSigmaR2()
+{
```
Comment (andy31415) [inline src/transport/CASESession.cpp:412]: This is ~150 lines, with 15 local variables on top and an extra 5 variables within scopes. 

Could we split things up and maybe add more focused unit tests for various steps? I find hard to review and maintain such large pieces of code.

File: src/credentials/tests/TestChipOpCred.cpp [hunk 18]
```diff
@@ -0,0 +1,287 @@
...
+static const BitFlags<uint16_t, KeyUsageFlags> sDSandEO(sDS.Raw() | sEO.Raw());
+static const BitFlags<uint16_t, KeyUsageFlags> sDSandDO(sDS.Raw() | sDO.Raw());
+static const BitFlags<uint16_t, KeyUsageFlags> sKCandDS(sKC.Raw() | sDS.Raw());
+static const BitFlags<uint16_t, KeyUsageFlags> sKCandNR(sKC.Raw() | sNR.Raw());
+static const BitFlags<uint16_t, KeyUsageFlags> sKCandKE(sKC.Raw() | sKE.Raw());
+static const BitFlags<uint16_t, KeyUsageFlags> sKCandDE(sKC.Raw() | sDE.Raw());
+static const BitFlags<uint16_t, KeyUsageFlags> sKCandKA(sKC.Raw() | sKA.Raw());
+static const BitFlags<uint16_t, KeyUsageFlags> sKCandCR(sKC.Raw() | sCR.Raw());
+static const BitFlags<uint16_t, KeyUsageFlags> sKCandEO(sKC.Raw() | sEO.Raw());
+static const BitFlags<uint16_t, KeyUsageFlags> sKCandDO(sKC.Raw() | sDO.Raw());
```
Comment (andy31415) [inline src/credentials/tests/TestChipOpCred.cpp:102]: Do we need these many bitflags constants?

I believe bitflags were built to keep the flags typesafe and they look a bit odd as constant values.
I would prefer to use KeyUsageFlags, KeyPurposeFlags, TestCertLoadFlags and so on as theya are. Re-naming the purpose flags with new names seems redundant.

Comment (andy31415) [reply src/credentials/tests/TestChipOpCred.cpp:102]: Where are these used? I did a search for some random ones (the first in the list above: sDSS and sDSandNR) and could not find any usage.

**Review by andy31415 — CHANGES_REQUESTED** [review]
Marking request changes: this needs some code cleanup/simplificatioon to be easier too maintain.

File: src/credentials/CHIPOpCred.h (hunk 10 above)
Comment (jpk233) [reply src/credentials/CHIPOpCred.h:168]: Yes, the initial idea is for the caller of the CASE Session to own the entire set of credentials, in order to avoid high amounts of memory copy and allocations.

File: src/transport/CASESession.h (hunk 11 above)
Comment (jpk233) [reply src/transport/CASESession.h:224]: The hash SHA256 stream is used to compute the salt for generating sigmaR2/3 and session Keys.
It is used to keep track of the previous Sigma messages that will be out of scope within the following Sigma message handling.
What you suggest is also an option, but will require us to store the sigma messages within the class's scope.

File: src/transport/CASESession.h (hunk 12 above)
Comment (jpk233) [reply src/transport/CASESession.h:213]: Do we have such mechanism in CHIP repo? To obtain current date and time?

File: src/transport/CASESession.h (hunk 13 above)
Comment (jpk233) [reply src/transport/CASESession.h:59]: I guess the idea here is to fix all the class' data types to a certain fixed length value in bytes. This is just mirroring what has been done in PASESession (with bool types).
Yes, I agree replacing uint64_t with NodeID for clarity

File: src/credentials/CHIPOpCred.cpp (hunk 3 above)
Comment (jpk233) [reply src/credentials/CHIPOpCred.cpp:30]: Yes, that's something we can do.

File: src/credentials/CHIPOpCred.h [hunk 19]
```diff
@@ -0,0 +1,184 @@
...
+ *
+ *  @brief
+ *    Collection of CHIPCertificateSet providing methods for
+ *    finding and validating Device Credentials against Roof of Trust chains.
+ */
+class DLL_EXPORT ChipOperationalCredentialSet
+{
+public:
+    ChipOperationalCredentialSet();
+    ~ChipOperationalCredentialSet();
```
[the remark this answers is not in the document]
Comment (jpk233) [reply src/credentials/CHIPOpCred.h:48]: Yes, can do.

File: src/credentials/CHIPOpCred.h (hunk 4 above)
Comment (jpk233) [reply src/credentials/CHIPOpCred.h:81]: Will do so.

**Comment by jpk233** [comment]
Continuing to work on this

**Comment by bzbarsky-apple** [comment]
Note: https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/2010 is in the middle of changing the payload of CASE messages in the spec.  This PR might want to either wait on that or be followed up with something to align to the spec changes.
